src/include/network/connection_handler_task.h

```diff
@@ -0,0 +1,87 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// network_handler_task.h
+//
+// Identification: src/include/network/network_handler_task.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <event2/buffer.h>
+#include <event2/bufferevent.h>
+#include <event2/event.h>
+#include <event2/listener.h>
+
+#include <unistd.h>
+
+#include "common/exception.h"
+#include "common/logger.h"
+#include "container/lock_free_queue.h"
+#include "network/notifiable_task.h"
+
+// TODO(tianyu) Do we really want to hard code this?
+#define QUEUE_SIZE 100
+
+namespace peloton {
+namespace network {
+
+/**
+ * A ConnectionHandlerTask is responsible for interacting with a client
+ * connection.
+ *
+ * A client connection, once taken by the dispatch, is sent to a handler. Then
+ * all related
+ * client events are registered in the handler task. All client interaction
+ * happens on the
+ * same ConnectionHandlerTask thread for the enitre lifetime of the connection.
+ */
+class ConnectionHandlerTask : public NotifiableTask {
+ public:
+  /**
+   * Constructs a new ConnectionHandlerTask instance.
+   * @param task_id task_id a unique id assigned to this task.
+   */
+  explicit ConnectionHandlerTask(int task_id);
+
+  /**
+   * @brief Notifies this ConnectionHandlerTask that a new client connection
+   * should be handled at socket fd.
+   *
+   * This method is meant to be invoked on another thread (probably the
+   * dispatcher) to update
+   * the necessary data structure so the handler thread is woken up.
+   *
+   * @param conn_fd the client connection socket fd.
+   */
+  void Notify(int conn_fd);
+
+  /**
+   * @brief Handles a new client assigned to this handler by the dispatcher.
+   *
+   * This method will create the necessary data structure for the client and
+   * register its event base to receive updates with appropriate callbacks
+   * when the client writes to the socket.
+   *
+   * @param new_conn_recv_fd the socket fd of the new connection
+   * @param flags unused. For compliance with libevent callback interface.
+   */
+  void HandleDispatch(int new_conn_recv_fd, short flags);
+
+ private:
+  // Notify new connection pipe(send end)
+  int new_conn_send_fd_;
+
+  // TODO(tianyu) Do we really need this queue, now that we are only sending an
+  // int?
+  // The queue for new connection requests, elements represent the socket fd of
+  // new client connections
+  LockFreeQueue<int> new_conn_queue_;
```

Comment:
We need this queue to pass the fd. This is relatively scalable compared to using some other shared data structure owned by the main thread.

Reply:
The question is, if we are just sending an int, why can't we just send that int though the fd we wrote 'c' to? I can understand that serializing more complex data structures is problematic, but for int this seems unnecessary.